### PR TITLE
bpo-36290: Fix kwargs handling in ast node constructors

### PR DESCRIPTION
**Pull request by remilapeyre**
https://bugs.python.org/issue36290

### CODE REVIEW

**Review by auvipy — APPROVED** [review]

File: Parser/asdl_c.py [hunk 1]
```diff
@@ -687,19 +690,50 @@ def visitModule(self, mod):
...
     }
     if (kw) {
         i = 0;  /* needed by PyDict_Next */
+        if (_PyObject_LookupAttrId(fields, &PyId_index, &index) < 0) {
+            goto cleanup;
+        }
         while (PyDict_Next(kw, &i, &key, &value)) {
+            PyObject* pos = _PyObject_FastCallDict(index, &key, 1, NULL);
+            PyObject* err = PyErr_Occurred();
+            if (pos == NULL || err != NULL) {
```
Comment (ZackerySpytz) [inline Parser/asdl_c.py]: Just check for `pos == NULL`. An exception will be set if `_PyObject_FastCallDict()` returned `NULL`.

Comment (matrixise) [reply Parser/asdl_c.py]: Hi @remilapeyre Do you have time for a fix in your PR?
Thank you

**Comment by remilapeyre** [comment]
@maxtrixise, sorry I missed Zackery comment, I will fix this when I get home.

**Comment by matrixise** [comment]
@remilapeyre Nice, I wanted to check all of your PRs.

**Comment by remilapeyre** [comment]
Hi @matrixise, I've made the requested changes and resolved the conflicts, it should be ready for a new review :)

File: Parser/asdl_c.py [hunk 2]
```diff
@@ -690,19 +695,50 @@ def visitModule(self, mod):
...
+                if (!PyErr_GivenExceptionMatches(err, PyExc_ValueError)) {
+                    res = -1;
+                    goto cleanup;
+                }
+                PyErr_Clear();
+            }
+            else {
+                Py_ssize_t p = PyNumber_AsSsize_t(pos, NULL);
+                Py_DECREF(pos);
+                if (p == -1 && PyErr_Occurred()) {
```
Comment (isidentical) [inline Parser/asdl_c.py]: Why not `p == -1 || PyErr_Occurred()`?

File: Lib/ast.py [hunk 3]
```diff
@@ -503,6 +503,14 @@ def __instancecheck__(cls, inst):
         return type.__instancecheck__(cls, inst)
 
 def _new(cls, *args, **kwargs):
+    for key in kwargs:
+        try:
+            pos = cls._fields.index(key)
+            if pos < len(args):
+                raise TypeError(f"{cls.__name__} got multiple values for argument {key!r}")
+        except ValueError:
+            # arbitrary keywork arguments are accepted
```
Comment (isidentical) [inline Lib/ast.py]:
```suggestion
            # arbitrary keyword arguments are accepted 
```

File: Lib/test/test_ast.py [hunk 4]
```diff
@@ -378,6 +378,8 @@ def test_classattrs(self):
 
         self.assertRaises(TypeError, ast.Num, 1, None, 2)
         self.assertRaises(TypeError, ast.Num, 1, None, 2, lineno=0)
+        self.assertRaises(TypeError, ast.Num, 1, n=2)
+        self.assertRaises(TypeError, ast.Constant, 1, value=2)
```
Comment (isidentical) [inline Lib/test/test_ast.py]: I think you should test some more examples with matching the message you are throwing (classname got multiple values for key etc.)

File: Misc/NEWS.d/next/Library/2019-03-17-19-01-53.bpo-36290.7VXo_K.rst [hunk 5]
```diff
@@ -0,0 +1,2 @@
+_ast.ast_type_init now raises TypeError on conflicting keyword arguments.
```
Comment (isidentical) [inline Misc/NEWS.d/next/Library/2019-03-17-19-01-53.bpo-36290.7VXo_K.rst]:
```suggestion
AST nodes are now raising :exc:`TypeError` on conflicting keyword arguments.
```
could be more descriptive

**Comment by remilapeyre** [comment]
Hi @isidentical, thanks for the review, it should be good now.

File: Parser/asdl_c.py [hunk 6]
```diff
@@ -628,6 +628,9 @@ class PyTypesVisitor(PickleVisitor):
 
     def visitModule(self, mod):
         self.emit("""
+_Py_IDENTIFIER(_fields);
+_Py_IDENTIFIER(_attributes);
+_Py_IDENTIFIER(index);
```
Comment (isidentical) [inline Parser/asdl_c.py]: If you look out to other usages for methods, you can see that we use `astmodulestate_global` to keep that references. For consistency with other parts of program, IMHO you should add these to there. (Don't repeat things like `_fields` that already present in state)

File: Parser/asdl_c.py [hunk 7]
```diff
@@ -690,19 +695,50 @@ def visitModule(self, mod):
...
-        if (res < 0)
+        if (res < 0) {
             goto cleanup;
+        }
     }
     if (kw) {
         i = 0;  /* needed by PyDict_Next */
+        if (_PyObject_LookupAttrId(fields, &PyId_index, &index) < 0) {
+            goto cleanup;
+        }
```
Comment (isidentical) [inline Parser/asdl_c.py]: Is there a reason that why you didn't use `PySequence_Index` and manually accessed it?

File: Parser/asdl_c.py (hunk 6 above)
Comment (remilapeyre) [reply Parser/asdl_c.py]: Thanks, I removed them.

File: Parser/asdl_c.py (hunk 7 above)
Comment (remilapeyre) [reply Parser/asdl_c.py]: I didn't know `PySequence_Index` existed, TIL.

**Review by isidentical — COMMENTED** [review]
It looks OK (except a few minor nits) but I wonder why you choose EAFP rather then LBYL for this case. I guess you can just do a contains check, no? and it would look simpler then the current version.By the way thanks for your active work on PR.

**Comment by isidentical** [comment]
> I'm not sure what you mean in this context, no exception is raised so I don't see how I could ask for forgiveness. 

Sorry about my conversation skills, I'm still trying to improve them. I mean that instead of that `try/except` for `ValueError` that is raised from `.index` when the value doesn't present, you could just use a contains check which would look more strict then just capturing raw `ValueError`s.

```py
def _new(cls, *args, **kwargs):
    for key in kwargs:
        if key not in cls._fields:
            continue # arbitrary keyword arguments are accepted        
        pos = cls._fields.index(key)
        if pos < len(args):
            raise TypeError(f"{cls.__name__} got multiple values "
                            f"for argument {key!r}")
    if cls in _const_types:
        return Constant(*args, **kwargs)
    return Constant.__new__(cls, *args, **kwargs)
```

Something like this is what I meant (of course also in the C implementation). What do you think about this?

**Comment by remilapeyre** [comment]
Ok, I agree it make the Python code better but for the C version all return codes need to be checked anyway so I'm not sure it's better:

```c

        i = 0;  /* needed by PyDict_Next */
        while (PyDict_Next(kw, &i, &key, &value)) {
            int contains = PySequence_Contains(fields, key);
            if (contains == -1) {
                res = -1;
                goto cleanup;
            } else if (contains == 0) {
                continue;
            }
            Py_ssize_t p = PySequence_Index(fields, key);
            if (p == -1) {
                res = -1;
                goto cleanup;
            }
            else {
                if (p < PyTuple_GET_SIZE(args)) {
                    PyErr_Format(PyExc_TypeError,
                        "%.400s got multiple values for argument '%U'",
                        Py_TYPE(self)->tp_name, key);
                    res = -1;
                    goto cleanup;
                }
            }
            res = PyObject_SetAttr(self, key, value);
            if (res < 0) {
                goto cleanup;
            }
        }
```

**Comment by isidentical** [comment]
I dont think it is good idea to assume ValueError only comes when the item is not present, it might come from something else in the future (unlikely, but why dont we make this more strict if we have a chance).

**Comment by remilapeyre** [comment]
That makes sense, here's a new patch without this issue.

**Review by isidentical — APPROVED** [review]
Looks great, it might worth to consider to add a case about arbitrary keyword arguments (both pseudo compatibility nodes and real nodes)

**Review by pablogsal — APPROVED** [review]
LGTM

**Comment by pablogsal** [comment]
Thanks for the PR @remilapeyre and thanks for the review @isidentical !

**Comment by isidentical** [comment]
Thank you for your continous work on this PR @remilapeyre and thanks for the merge @pablogsal :confetti_ball: :partying_face:

**Pull request merged**